From: ossia SDK maintainers
Date: Thu, 27 Aug 2026 00:00:00 +0000
Subject: [PATCH] llvm: adapt to LLVM 23 APIs

LLVM 23 removed the legacy fast-math fields from TargetOptions, changed the
single-argument target lookup to take a Triple, and generalized
CreateUnaryIntrinsic's return type to Value because it may fold the call.
---
diff --git a/compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp b/compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp
index 9c2d72e..6bc0bf6 100644
--- a/compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp
+++ b/compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp
@@ -312,13 +312,15 @@ bool llvm_dynamic_dsp_factory_aux::initJIT(string& error_msg)
 #if LLVM_VERSION_MAJOR < 22
         targetOptions.UnsafeFPMath = true;
 #endif
-        targetOptions.NoInfsFPMath          = true;
-        targetOptions.NoNaNsFPMath          = true;
+#if LLVM_VERSION_MAJOR < 23
+        targetOptions.NoInfsFPMath = true;
+        targetOptions.NoNaNsFPMath = true;
+#endif
         targetOptions.GuaranteedTailCallOpt = true;
         targetOptions.NoTrappingFPMath      = true;
     }
 
-#if LLVM_VERSION_MAJOR >= 9
+#if LLVM_VERSION_MAJOR >= 9 && LLVM_VERSION_MAJOR < 23
     targetOptions.NoSignedZerosFPMath = true;
 #endif
 
@@ -500,7 +502,11 @@ bool llvm_dynamic_dsp_factory_aux::writeDSPFactoryToObjectcodeFileAux(
 #endif
 
     string Error;
-    auto   Target = TargetRegistry::lookupTarget(TargetTriple, Error);
+#if LLVM_VERSION_MAJOR >= 23
+    auto Target = TargetRegistry::lookupTarget(Triple(TargetTriple), Error);
+#else
+    auto Target = TargetRegistry::lookupTarget(TargetTriple, Error);
+#endif
 
     // Print an error and exit if we couldn't find the requested target.
     // This generally occurs if we've forgotten to initialise the
diff --git a/compiler/generator/llvm/llvm_instructions.hh b/compiler/generator/llvm/llvm_instructions.hh
index 80dadce..6333e65 100644
--- a/compiler/generator/llvm/llvm_instructions.hh
+++ b/compiler/generator/llvm/llvm_instructions.hh
@@ -975,11 +975,13 @@ class LLVMInstVisitor : public InstVisitor, public LLVMTypeHelper {
 #if LLVM_VERSION_MAJOR >= 8
             // LLVM unary intrinsic
         } else if (fUnaryIntrinsicTable.find(inst->fName) != fUnaryIntrinsicTable.end()) {
-            llvm::CallInst* call_inst =
+            llvm::Value* value_inst =
                 fBuilder->CreateUnaryIntrinsic(fUnaryIntrinsicTable[inst->fName], fun_args[0]);
-            AddAttributeAtIndex(call_inst, llvm::AttributeList::FunctionIndex,
-                                llvm::Attribute::Builtin);
-            fCurValue = call_inst;
+            if (auto* call_inst = llvm::dyn_cast<llvm::CallInst>(value_inst)) {
+                AddAttributeAtIndex(call_inst, llvm::AttributeList::FunctionIndex,
+                                    llvm::Attribute::Builtin);
+            }
+            fCurValue = value_inst;
 
             // LLVM binary intrinsic
         } else if (fBinaryIntrinsicTable.find(inst->fName) != fBinaryIntrinsicTable.end()) {
